kafkazk/zookeeper_objects.go

```diff
@@ -0,0 +1,129 @@
+package kafkazk
+
+// TopicStateISR is a map of partition numbers to PartitionState.
+type TopicStateISR map[string]PartitionState
+
+// PartitionState is used for unmarshalling json data from a partition state:
+// e.g. /brokers/topics/some-topic/partitions/0/state
+type PartitionState struct {
+	Version         int   `json:"version"`
+	ControllerEpoch int   `json:"controller_epoch"`
+	Leader          int   `json:"leader"`
+	LeaderEpoch     int   `json:"leader_epoch"`
+	ISR             []int `json:"isr"`
+}
+
+// Reassignments is a map of topic:partition:brokers.
+type Reassignments map[string]map[int][]int
+
+// reassignPartitions is used for unmarshalling /admin/reassign_partitions data.
+type reassignPartitions struct {
+	Partitions []reassignConfig `json:"partitions"`
+}
+
+type reassignConfig struct {
+	Topic     string `json:"topic"`
+	Partition int    `json:"partition"`
+	Replicas  []int  `json:"replicas"`
+}
+
+// TopicConfig is used for unmarshalling  /config/topics/<topic> from ZooKeeper.
+type TopicConfig struct {
+	Version int               `json:"version"`
+	Config  map[string]string `json:"config"`
+}
+
+// TopicMetadata holds the topic data found in the /brokers/topics/<topic> znode.
+// This is designed for the version 3 fields present in Kafka version ~2.4+.
+type TopicMetadata struct {
+	Version          int
+	Name             string
+	TopicID          string `json:"topic_id"`
+	Partitions       map[int][]int
+	AddingReplicas   map[int][]int `json:"adding_replicas"`
+	RemovingReplicas map[int][]int `json:"removing_replicas"`
+}
+
+// KafkaConfig is used to issue configuration updates to either
+// topics or brokers in ZooKeeper.
+type KafkaConfig struct {
+	Type    string          // Topic or broker.
+	Name    string          // Entity name.
+	Configs []KafkaConfigKV // Config KVs.
+}
+
+// KafkaConfigKV is a [2]string{key, value} representing a Kafka configuration.
+type KafkaConfigKV [2]string
+
+// KafkaConfigData is used for unmarshalling /config/<type>/<name> data
+// from ZooKeeper.
+type KafkaConfigData struct {
+	Version int               `json:"version"`
+	Config  map[string]string `json:"config"`
+}
+
+// NewKafkaConfigData creates a KafkaConfigData.
+func NewKafkaConfigData() KafkaConfigData {
+	return KafkaConfigData{
+		Config: make(map[string]string),
+	}
+}
+
+// Reassignments returns a Reassignments from a given topics TopicMetadata.
+func (tm TopicMetadata) Reassignments() Reassignments {
```

Comment:
Should we have a Reassignment struct that contains the reassignment for only one topic? It's sort of weird for TopicMetadata to hold a kv struct where the only key is the current topic name.

Reply:
That's because `Reassignments` is a pre-existing type; this method is to translate the new type `TopicMetadata` (which belongs to a single topic) to a `Reassignments`. In the `ListReassignments` call, we get a `TopicMetadata`->`Reassignments` for all topics, merge them into a single `Reassignments`.